### PR TITLE
Support resources of selector in VApp and optimize protocol API

### PR DESCRIPTION
**Pull request by FillZpp**
Signed-off-by: FillZpp <FillZpp.pub@gmail.com>

Support resources of selector in VApp and optimize protocol API.

VirtualApp changes:

1. Add optional `resources` field in `MatchLimitSelector`, so that it can not only used by objectSelector(TODO), but also namespaceSelector/namespaceRegex.
2. Add optional `subsetPublicResources` field in `VirtualAppRoute`, so that users can specify public resources that can be shared and operated by all subsets (e.g. events).

Protobuf changes (a little big):

1. Mark function and messages as `v1`, to make it compatible if we have to refactor the API in future.
2. Remove `SpecHash`, which is too complicated. Just a `hash` in meta and it is only cared by controller, proxy has no need to calculate it.
3. Re-define fields in `RouteV1`. `globalLimits` and `subsetLimits` are more clear than previous definition.

GrpcClient in proxy changes:

1. Optimize `snapshot`, which contains protocol spec and can be renewed if possible.

### CODE REVIEW

File: util/hash.go [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+package util
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+)
+
+func GetMD5Hash(text string) string {
+	hash := md5.Sum([]byte(text))
```
[the remark this answers is not in the document]
Comment (FillZpp) [reply util/hash.go:25]: ignore

[the remark this answers is not in the document]
Comment (FillZpp) [reply util/hash.go:25]: ignore

File: apis/ctrlmesh/proto/ctrlmesh_internal.go [hunk 2]
```diff
@@ -17,117 +17,146 @@ limitations under the License.
...
+type InternalSubsetLimit struct {
+	Subset string
+	Limits []*InternalMatchLimitRule
+}
+
+type InternalMatchLimitRule struct {
+	Namespaces sets.String
+	// TODO: parse from objectSelector string in protobuf
+	// ObjectSelector    *metav1.LabelSelector
+	Resources []*APIGroupResourceV1
```
Comment (Eikykun) [inline apis/ctrlmesh/proto/ctrlmesh_internal.go:86]: For a special `Subset`. When we want to customize a rule to limit some resources(e.g. `v1.Pod`), and limit other resources with a diff rule. Should i specify all other resources in `APIGroupResource`?

Comment (FillZpp) [reply apis/ctrlmesh/proto/ctrlmesh_internal.go:86]: In a rule you can write multiple `match` and each of them can have its specific resources.

For example:

```yaml
  route:
    subRules:
    - name: canary-rule
      match:
      - namespaceSelector: # or objectSelector
          # ...
        resources:
          # v1.Pod
      - namespaceSelector: # or objectSelector
          # ...
        resources:
          # other resources
```

**Review by Eikykun — APPROVED** [review]

**Pull request merged**